crates/sui-json-rpc/src/authority_state.rs

```diff
@@ -0,0 +1,654 @@
+    ) -> StateReadResult<VerifiedCheckpoint>;
+
+    fn multi_get_executed_transactions(
+        &self,
+        digests: &[TransactionDigest],
+    ) -> StateReadResult<Vec<Option<VerifiedTransaction>>>;
+
+    fn multi_get_executed_effects(
+        &self,
+        digests: &[TransactionDigest],
+    ) -> StateReadResult<Vec<Option<TransactionEffects>>>;
+
+    fn deprecated_multi_get_transaction_checkpoint(
+        &self,
+        digests: &[TransactionDigest],
+    ) -> StateReadResult<Vec<Option<(EpochId, CheckpointSequenceNumber)>>>;
+
+    fn deprecated_get_transaction_checkpoint(
+        &self,
+        digest: &TransactionDigest,
+    ) -> StateReadResult<Option<(EpochId, CheckpointSequenceNumber)>>;
+
+    fn multi_get_checkpoint_by_sequence_number(
+        &self,
+        sequence_numbers: &[CheckpointSequenceNumber],
+    ) -> StateReadResult<Vec<Option<VerifiedCheckpoint>>>;
+
+    fn multi_get_events(
+        &self,
+        digests: &[TransactionEventsDigest],
+    ) -> StateReadResult<Vec<Option<TransactionEvents>>>;
+
+    #[allow(clippy::type_complexity)]
+    async fn dry_exec_transaction(
+        &self,
+        transaction: TransactionData,
+        transaction_digest: TransactionDigest,
+    ) -> StateReadResult<(
+        DryRunTransactionBlockResponse,
+        BTreeMap<ObjectID, (ObjectRef, Object, WriteKind)>,
+        TransactionEffects,
+        Option<ObjectID>,
+    )>;
+    async fn dev_inspect_transaction_block(
+        &self,
+        sender: SuiAddress,
+        transaction_kind: TransactionKind,
+        gas_price: Option<u64>,
+    ) -> StateReadResult<DevInspectResults>;
+
+    fn get_total_transaction_blocks(&self) -> StateReadResult<u64>;
+    async fn get_transaction_block(
+        &self,
+        digest: TransactionDigest,
+    ) -> StateReadResult<VerifiedTransaction>;
+
+    fn get_executed_effects(
+        &self,
+        digest: TransactionDigest,
+    ) -> StateReadResult<TransactionEffects>;
+    fn get_checkpoint_by_sequence_number(
+        &self,
+        sequence_number: CheckpointSequenceNumber,
+    ) -> StateReadResult<Option<VerifiedCheckpoint>>;
+
+    fn get_transaction_events(
+        &self,
+        digest: &TransactionEventsDigest,
+    ) -> StateReadResult<TransactionEvents>;
+
+    fn get_latest_checkpoint_sequence_number(&self) -> StateReadResult<CheckpointSequenceNumber>;
+    fn get_checkpoints(
+        &self,
+        // If `Some`, the query will start from the next item after the specified cursor
+        cursor: Option<CheckpointSequenceNumber>,
+        limit: u64,
+        descending_order: bool,
+    ) -> StateReadResult<Vec<Checkpoint>>;
+
+    fn loaded_child_object_versions(
+        &self,
+        transaction_digest: &TransactionDigest,
+    ) -> StateReadResult<Option<Vec<(ObjectID, SequenceNumber)>>>;
+
+    fn get_chain_identifier(&self) -> StateReadResult<ChainIdentifier>;
+}
+
+#[async_trait]
+impl State for AuthorityState {
```

Comment:
Do we expect to add any other implementations for this trait? The super lengthy traits can get quite unwieldy to implement again if needed for testing, etc., which can perhaps be addressed by breaking this up into a few separate traits each with related functions.

Reply:
Ah I see, unfamiliar with rust so not sure if possible, but can chunk this into a trait that all APIs share and then maybe for each api can have a trait extending the base one

Reply:
If it is possible to separate the functions in this trait into a few smaller sub-traits, I would recommend that, but if this is the only instance of the trait that we expect to have, then it wouldn't really matter

Reply:
It's definitely possible to separate the functions out into smaller traits, but after trying this out, I think I'm inclined towards keeping this a single trait for now.
1. If we need additional authority state methods, we only need to add it to this trait and its impl. automock does automatic mocking for us. And so in a unit test, we only need to explicitly provide an implementation for behavior we want to test.
2. It's not clear rn whether to break up the functions by read path (kv_store, db, epoch_store, etc.) or by api.

Conversely, when broken into smaller sub-traits, the composition/ inheritance becomes harder to mock, and brings more work than benefits. But I can def revisit this if fetching data from state becomes more complex